# Data cleaning using pandas

In this notebook, you'll see how clean structured data using `pandas`. We are aware that you have already done an extensive review of pandas, and are now experts in data manipulation, so we'll keep it short.

We are going to use a dataset about New York City Squirrels taken from [NYC Open Data](https://opendata.cityofnewyork.us/) which is a great resource for data sets.
This dataset is quite small and this notebook will just be an overview of different stuff to look out for when cleaning data.


In [3]:
import pandas as pd
import numpy as np
filename="./2018_Central_Park_Squirrel_Census_-_Squirrel_Data.csv"
df = pd.read_csv(filename)
df.head()

X          Y Unique Squirrel ID Hectare Shift      Date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   Hectare Squirrel Number    Age Primary Fur Color Highlight Fur Color  ...  \
0                        3    NaN               NaN                 NaN  ...   
1                        4    NaN               NaN                 NaN  ...   
2                        8    NaN              Gray                 NaN  ...   
3                       14  Adult              Gray                 NaN  ...   
4                        5  Adult              Gray            Cinnamon  ...   

    Kuks  Quaas  Moans Tail flags Tail twitches  Approaches  Indifferent  \
0  False  False  False      False         False       False        False   
1  False  False  False      False         False       False        False   
2  False  False  False      False         False       False        False   
3  False  False  False      False         False       False        False   
4  False  False  False      False         False       False        False   

   Runs from  Other Interactions                                    Lat/Long  
0      False                 NaN  POINT (-73.9561344937861 40.7940823884086)  
1      False                 NaN  POINT (-73.9688574691102 40.7837825208444)  
2      False                 NaN  POINT (-73.97428114848522 40.775533619083)  
3       True                 NaN  POINT (-73.9596413903948 40.7903128889029)  
4      False                 NaN  POINT (-73.9702676472613 40.7762126854894)  

[5 rows x 31 columns]


## Things to check
### Missing values. 

There are different ways that a value can be missing. It can be just an empty string (`""`), or `NaN` (Not a Number).
`np.nan`, `None` and `NaT` (for `datetime64` types) are standard missing value for Pandas.

You can either delete the rows/attribute if there are too many missing, or replace them. I can't give you a certain threshold to delete or replace, it really depends on the variable, on the context, and what you are going to do with the data later on.
In the same vein, there are many different ways to replace a value, and it will depend on the type of data you have. [[More info here]](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html)

Pandas provides `isnull()`, `isna()` functions to detect missing values. Both of them do the same thing.

In [4]:
print("There are {} rows of data".format(len(df)))

There are 3023 rows of data


We see that we have 3023 rows of data. Looking at the number of NaN values per column, one can confidently say that columns where we have > 1000 missing values are not worth keeping. The case can be made that, for certain scenarios, the columns `Highlight Fur Color` should be kept too, but here we are going to drop it because the column `Combination of Primary and Highlight Color` combines both. 

This notebook is going to feature small code blocks that you'll have to complete.

The first task is to count how many nan values there are per attribute.  
Hint: use `df.isnull()`

In [15]:
# Checking how many rows of each attribute are NaN
df.isnull().sum()

X                                                0
Y                                                0
Unique Squirrel ID                               0
Hectare                                          0
Shift                                            0
Date                                             0
Hectare Squirrel Number                          0
Age                                            121
Primary Fur Color                               55
Highlight Fur Color                           1086
Combination of Primary and Highlight Color       0
Color notes                                   2841
Location                                        64
Above Ground Sighter Measurement               114
Specific Location                             2547
Running                                          0
Chasing                                          0
Climbing                                         0
Eating                                           0
Foraging                       

Now that we've identified all the columns that have NaN values, we would like to drop those with more than 1000 NaN values.
Can you do that ?  
Hint: `df.drop()`  
Moreover, can you print the columns that were dropped, along with the number of missing values that they had ?

In [29]:
# Drop the columns that have more than 1000 missing values
df.dropna(thresh=1000)
        
# Print the columns that were dropped


<ipython-input-29-99114a5a65e5>:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df.isnull().sum()>1000]


IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).

### Attribute type

Sometimes, an attribute does not have the correct type, e.g. integers are labeled as floats. So it is useful to check if it needs to be changed or not.


Please print the dtypes of every attribute.

In [32]:
#Print dtype of every attribute
df.dtypes

X                                             float64
Y                                             float64
Unique Squirrel ID                             object
Hectare                                        object
Shift                                          object
Date                                            int64
Hectare Squirrel Number                         int64
Age                                            object
Primary Fur Color                              object
Highlight Fur Color                            object
Combination of Primary and Highlight Color     object
Color notes                                    object
Location                                       object
Above Ground Sighter Measurement               object
Specific Location                              object
Running                                          bool
Chasing                                          bool
Climbing                                         bool
Eating                      

For example, for `Hectare`, one could think that the variable is just the size of the plot of land. However, it seems it is just an identifier for which plot of land it is. One would have to see the documentation to know for sure. In this case, the `object` dtype is appropriate.

In [34]:
# Print unique values of the 'Hectare' attribute
df.Hectare.unique()

array(['37F', '21B', '11B', '32E', '13E', '11H', '36H', '33F', '21C',
       '11D', '20B', '22F', '36I', '05C', '07H', '16C', '14E', '32A',
       '17F', '16I', '12I', '32F', '25A', '15E', '39G', '29I', '07E',
       '17C', '10A', '28A', '12B', '18A', '33H', '29C', '38C', '31H',
       '08H', '34F', '13H', '35C', '17E', '02B', '19D', '40B', '03E',
       '39C', '06G', '14F', '06I', '15F', '08B', '03A', '05E', '05F',
       '16E', '13F', '29H', '23C', '33C', '38G', '16G', '35A', '28C',
       '32D', '22A', '19C', '15G', '10F', '42E', '22B', '05A', '32H',
       '42C', '41D', '21G', '07F', '38F', '41B', '36A', '31A', '42G',
       '37D', '37E', '24B', '07G', '15I', '19B', '13D', '35B', '40D',
       '38B', '40I', '07B', '40A', '03B', '21D', '10B', '41A', '02F',
       '26A', '35E', '11G', '07A', '03D', '35D', '18G', '06F', '09F',
       '38E', '33G', '32G', '04D', '06H', '15D', '39H', '20F', '09G',
       '05D', '23F', '39A', '13A', '10G', '13C', '32C', '31D', '18C',
       '40E', '04G',

### Attribute values

Carefully check each attribute to see if the unique values are coherent and make sense.
Moreover, check that minimum and maximum values make sense. (e.g. age of a person that is negative)

Hint: use your previously acquired knowledge of extracting unique values.
Be careful to limit the attributes that you print by the number of unique values they have, as there are fields that have all their values that are unique, and we don't want a kilometer long list of each value.

In [60]:
# Print the number of unique values of each column, only printing those that have less than 100 unique values
column_list = df.columns.values.tolist()
print(column_list)
for column_name in column_list:
    if df[column_name].nunique()<100:
        print(column_name)
        print("-----------")
        print(df[column_name].unique())

['X', 'Y', 'Unique Squirrel ID', 'Hectare', 'Shift', 'Date', 'Hectare Squirrel Number', 'Age', 'Primary Fur Color', 'Highlight Fur Color', 'Combination of Primary and Highlight Color', 'Color notes', 'Location', 'Above Ground Sighter Measurement', 'Specific Location', 'Running', 'Chasing', 'Climbing', 'Eating', 'Foraging', 'Other Activities', 'Kuks', 'Quaas', 'Moans', 'Tail flags', 'Tail twitches', 'Approaches', 'Indifferent', 'Runs from', 'Other Interactions', 'Lat/Long']
Shift
-----------
['PM' 'AM']
Date
-----------
[10142018 10192018 10172018 10102018 10082018 10062018 10132018 10072018
 10182018 10122018 10202018]
Hectare Squirrel Number
-----------
[ 3  4  8 14  5  2  1  6  9 23  7 13 10 16 11 12 20 15 19 17 21 22 18]
Age
-----------
[nan 'Adult' 'Juvenile' '?']
Primary Fur Color
-----------
[nan 'Gray' 'Cinnamon' 'Black']
Highlight Fur Color
-----------
[nan 'Cinnamon' 'White' 'Gray' 'Cinnamon, White' 'Gray, White'
 'Black, Cinnamon, White' 'Black' 'Black, White' 'Black, Cinnamo

In [62]:
df.describe(include='all').T # Transpose the data frame so that it fits in a cell

count unique  \
X                                           3023    NaN   
Y                                           3023    NaN   
Unique Squirrel ID                          3023   3018   
Hectare                                     3023    339   
Shift                                       3023      2   
Date                                        3023    NaN   
Hectare Squirrel Number                     3023    NaN   
Age                                         2902      3   
Primary Fur Color                           2968      3   
Highlight Fur Color                         1937     10   
Combination of Primary and Highlight Color  3023     22   
Color notes                                  182    135   
Location                                    2959      2   
Above Ground Sighter Measurement            2909     41   
Specific Location                            476    304   
Running                                     3023      2   
Chasing                                     3023      2   
Climbing                                    3023      2   
Eating                                      3023      2   
Foraging                                    3023      2   
Other Activities                             437    307   
Kuks                                        3023      2   
Quaas                                       3023      2   
Moans                                       3023      2   
Tail flags                                  3023      2   
Tail twitches                               3023      2   
Approaches                                  3023      2   
Indifferent                                 3023      2   
Runs from                                   3023      2   
Other Interactions                           240    197   
Lat/Long                                    3023   3023   

                                                                                          top  \
X                                                                                         NaN   
Y                                                                                         NaN   
Unique Squirrel ID                                                             40B-AM-1019-06   
Hectare                                                                                   14D   
Shift                                                                                      PM   
Date                                                                                      NaN   
Hectare Squirrel Number                                                                   NaN   
Age                                                                                     Adult   
Primary Fur Color                                                                        Gray   
Highlight Fur Color                                                                  Cinnamon   
Combination of Primary and Highlight Color                                              Gray+   
Color notes                                 Gray & Cinnamon selected as Primary. White sel...   
Location                                                                         Ground Plane   
Above Ground Sighter Measurement                                                        FALSE   
Specific Location                                                                        tree   
Running                                                                                 False   
Chasing                                                                                 False   
Climbing                                                                                False   
Eating                                                                                  False   
Foraging                                                                                False   
Other Activities                                                                      digging   
Kuks                

**Does something stand out to you ?**

In the column `Combination of Primary and Highlight Color` for example, we have a value that is only `'+'`.
Again, there are multiple ways to deal with this. When you can't figure out what the value means (like in this case), you can drop it. Sometimes, when you are able to figure out which one is meant, you can replace it with the correct value. This is where the domain knowledge come into play, knowing which variable can be replaced with what, and what it could mean.

Another example, from another dataset containing vehicle types where the vehicle type was manually encoded.
You see how 'Tow truck' appears in many different forms.
The right approach is to consolidate all of the different values of tow truck into a single one.

In [63]:
trucks = ['Pick-up Truck', 'Tractor Truck Diesel', 'Box Truck', 'Tractor Truck Gasoline', 'Tow Truck / Wrecker', 'Beverage Truck', 'truck', 'Armored Truck', 'TRUCK', 'Truck', 'Tow Truck', 'FIRE TRUCK', 'PICK-UP TRUCK', 'BOX TRUCK', 'Fire Truck', 'box truck', 'TRUCK FLAT', 'FIRTRUCK', 'TRUCK VAN', 'FIRETRUCK', 'FDNY TRUCK', 'dump truck', 'Fire truck', 'UPS TRUCK', 'fire truck', 'FOOD TRUCK', 'MAIL TRUCK', 'USPS TRUCK', 'tow truck', 'TOW TRUCK', 'Tow truck']
[truck for truck in trucks if 'tow' in truck.lower()]

['Tow Truck / Wrecker', 'Tow Truck', 'tow truck', 'TOW TRUCK', 'Tow truck']

### Duplicates

Search for duplicates and remove them



In [64]:
# There are no duplicates
df.duplicated().any()

False

If there had been, call `df.drop_duplicates(inplace=True)` to remove them.

Again, this is just a short introduction to data cleaning of structured data, there are many more steps that one could optionally take.

## References and reading material

[Data cleaning challenge - Kaggle](https://www.kaggle.com/rtatman/data-cleaning-challenge-handling-missing-values)  
[Preparing data for a machine learning model - Dataquest](https://www.dataquest.io/blog/machine-learning-preparing-data/)  
[Preliminary data cleaning - Kaggle](https://www.kaggle.com/quannguyen135/preliminary-data-cleaning-with-python)  
[Pandas data cleaning cheatsheet - Elite Data Science](https://elitedatascience.com/python-cheat-sheet)